File: server/riskEngine.js
/**
 * Server-side Risk Scoring Engine implementing Section 5 rules.
 */
export function calculateRiskScore(claim) {
  let score = 0;
  const flags = [];

  const policyStartDate = new Date(claim.policyStartDate);
  const incidentDate = new Date(claim.incidentDate);
  const daysBetweenPolicyAndIncident = Math.floor((incidentDate - policyStartDate) / (1000 * 60 * 60 * 24));

  // Rule 1 & 2: Incident timing relative to policy start
  if (daysBetweenPolicyAndIncident < 30 && daysBetweenPolicyAndIncident >= 0) {
    score += 40;
    flags.push({
      flag: "possible waiting-period violation",
      impact: 40,
      severity: "alert",
      explanation: `Incident occurred ${daysBetweenPolicyAndIncident} days after policy start (under mandatory 30-day waiting threshold).`
    });
  } else if (daysBetweenPolicyAndIncident < 90 && daysBetweenPolicyAndIncident >= 30) {
    score += 15;
    flags.push({
      flag: "within early-claim window",
      impact: 15,
      severity: "warning",
      explanation: `Incident occurred ${daysBetweenPolicyAndIncident} days after policy inception (early claim window < 90 days).`
    });
  }

  // Rule 3 & 4: Claim amount vs Sum Insured
  const sumInsured = Number(claim.sumInsured) || 1;
  const claimAmount = Number(claim.claimAmount) || 0;
  const ratio = claimAmount / sumInsured;

  if (ratio > 0.90) {
    score += 30;
    flags.push({
      flag: "unusually high vs sum insured",
      impact: 30,
      severity: "alert",
      explanation: `Claim amount (₹${claimAmount.toLocaleString()}) represents ${(ratio * 100).toFixed(1)}% of total policy sum insured (₹${sumInsured.toLocaleString()}).`
    });
  } else if (ratio > 0.60) {
    score += 10;
    flags.push({
      flag: "substantial proportion of sum insured",
      impact: 10,
      severity: "warning",
      explanation: `Claim amount (₹${claimAmount.toLocaleString()}) represents ${(ratio * 100).toFixed(1)}% of sum insured (₹${sumInsured.toLocaleString()}).`
    });
  }

  // Rule 5: Supporting documents count
  const docCount = Array.isArray(claim.documents) ? claim.documents.length : 0;
  if (docCount < 2) {
    score += 15;
    flags.push({
      flag: "may be insufficient for verification",
      impact: 15,
      severity: "warning",
      explanation: `Only ${docCount} supporting document(s) provided. Minimum 2 required for standard automatic verification.`
    });
  }

  // Rule 6: Description brevity
  const wordCount = (claim.description || "").trim().split(/\s+/).filter(Boolean).length;
  if (wordCount < 12) {
    score += 10;
    flags.push({
      flag: "very brief, may need follow-up",
      impact: 10,
      severity: "warning",
      explanation: `Description contains only ${wordCount} words, which is below the recommended detail length.`
    });
  }

  // Cap total at 100
  const finalScore = Math.min(100, Math.max(0, score));

  // Determine band
  let band = "low"; // 0-19
  let color = "#3E6E5B"; // Sage
  if (finalScore >= 50) {
    band = "high";
    color = "#A6394A"; // Rust
  } else if (finalScore >= 20) {
    band = "medium";
    color = "#C8862A"; // Amber
  }

  return {
    riskScore: finalScore,
    riskBand: band,
    color,
    riskFlags: flags
  };
}

/**
 * Blend rule-based score with AWS Fraud Detector score (if present)
 */
export function blendFraudScore(ruleScore, fraudDetectorScore) {
  if (fraudDetectorScore == null || isNaN(fraudDetectorScore)) {
    return ruleScore;
  }
  // Weighted average: 60% rule score + 40% ML fraud detector score
  return Math.round((ruleScore * 0.6) + (fraudDetectorScore * 0.4));
}

File: backend/risk_engine.py
import math
from datetime import datetime

def calculate_risk_score(claim: dict) -> dict:
    score = 0
    flags = []

    try:
        policy_start = datetime.strptime(claim.get("policyStartDate", "2025-01-01"), "%Y-%m-%d")
        incident = datetime.strptime(claim.get("incidentDate", "2025-01-01"), "%Y-%m-%d")
        days = (incident - policy_start).days
    except Exception:
        days = 100

    # Rule 1 & 2: Incident timing relative to policy start
    if 0 <= days < 30:
        score += 40
        flags.append({
            "flag": "possible waiting-period violation",
            "impact": 40,
            "severity": "alert",
            "explanation": f"Incident occurred {days} days after policy start (under mandatory 30-day waiting threshold)."
        })
    elif 30 <= days < 90:
        score += 15
        flags.append({
            "flag": "within early-claim window",
            "impact": 15,
            "severity": "warning",
            "explanation": f"Incident occurred {days} days after policy inception (early claim window < 90 days)."
        })

    # Rule 3 & 4: Claim amount vs Sum Insured
    sum_insured = float(claim.get("sumInsured") or 1)
    claim_amount = float(claim.get("claimAmount") or 0)
    ratio = claim_amount / sum_insured if sum_insured > 0 else 0

    if ratio > 0.90:
        score += 30
        flags.append({
            "flag": "unusually high vs sum insured",
            "impact": 30,
            "severity": "alert",
            "explanation": f"Claim amount (₹{claim_amount:,.0f}) represents {(ratio * 100):.1f}% of total policy sum insured (₹{sum_insured:,.0f})."
        })
    elif ratio > 0.60:
        score += 10
        flags.append({
            "flag": "substantial proportion of sum insured",
            "impact": 10,
            "severity": "warning",
            "explanation": f"Claim amount (₹{claim_amount:,.0f}) represents {(ratio * 100):.1f}% of sum insured (₹{sum_insured:,.0f})."
        })

    # Rule 5: Supporting documents count
    docs = claim.get("documents") or []
    doc_count = len(docs) if isinstance(docs, list) else 0
    if doc_count < 2:
        score += 15
        flags.append({
            "flag": "may be insufficient for verification",
            "impact": 15,
            "severity": "warning",
            "explanation": f"Only {doc_count} supporting document(s) provided. Minimum 2 required for standard automatic verification."
        })

    # Rule 6: Description brevity
    desc = (claim.get("description") or "").strip()
    word_count = len([w for w in desc.split() if w])
    if word_count < 12:
        score += 10
        flags.append({
            "flag": "very brief, may need follow-up",
            "impact": 10,
            "severity": "warning",
            "explanation": f"Description contains only {word_count} words, which is below the recommended detail length."
        })

    final_score = min(100, max(0, score))
    band = "low"
    color = "#3E6E5B"
    if final_score >= 50:
        band = "high"
        color = "#A6394A"
    elif final_score >= 20:
        band = "medium"
        color = "#C8862A"

    return {
        "riskScore": final_score,
        "riskBand": band,
        "color": color,
        "riskFlags": flags
    }

def blend_fraud_score(rule_score: int, fraud_score: float = None) -> int:
    if fraud_score is None:
        return rule_score
    return round((rule_score * 0.6) + (fraud_score * 0.4))
